### ISSUE TITLE
Introduce with_named_binding

### ISSUE THREAD

**Issue by josevalim** [elixir-ecto/ecto#3970]
### Elixir version

1.14

### Database and Version

All

### Ecto Versions

All

### Database Adapter and Versions (postgrex, myxql, etc)

All

### Current behavior

Today we often have to write this code when a named binding may be set conditionally:

```elixir
  if has_named_binding?(query, :comments) do
    query
  else
    join(query, :left, c in assoc(p, :comments), as: :comments)
  end
```

Should we provide a small function that encapsulates this pattern?

### Expected behavior

One option is to introduce a `with_named_binding`:

```elixir
def with_named_binding(query, name, fun) do
  if has_named_binding?(query, name) do
    query
  else
    fun.(query)
  end
end
```

Then we can rewrite it as:

```elixir
  with_named_binding(query, :comments, fn ->
    join(query, :left, c in assoc(p, :comments), as: :comments)
  end)
```

Another option is to make it specific to joins and introduce a `with_named_join`:

```elixir
with_named_join(query, ..., as: ...)
```

We would only add the join to the query if the `:as` does not already exist. This is potentially less typing but also more magic.

Thoughts?

**Comment by josevalim** [comment]
You can only name `from` and `join` and if you have a query, the `from` has already been defined, so you can't add a `from` while checking for named bindings, if that makes any sense.

**Comment by greg-rychlewski** [comment]
Ah, I thought combinations are sharing bindings but was mistaken.

I personally never use `has_named_binding` but if it's almost always used to conditionally add a join then something like the second one makes more sense to me. It's more magic but it's clearer what the purpose of the function is.

The name might be able to be clearer as well. Something like `new_named_join` to convey the intent is to only add the join if it doesn't exist. And then I guess the arguments could be the same as the regular join except enforcing `:as` is specified. 

`with_name_binding` might already be clear to others though and I might be lacking context. But I was thinking of the `put_new` functions from `Map` and `Keyword`. And `with` makes me think "take the existing values, put them into this function and do something with it".

**Comment by AndrewDryga** [comment]
One of the solutions is something that is very common in our codebase, we for most of the schemas that have preloads we would have a `Query` module that will allow to preload some of assocs using joins (the ones that are faster to join rather than doing separate queries):
```
  def with_preloaded_something(queryable) do
    if has_named_binding?(queryable, :something) do
      queryable
    else
      queryable
      |> join(...)
      |> preload(...)
    end
  end
```

This might be a slightly different use-case than you had in mind though because your examples are focused around joins, while we actually do both joins and preloads all the time. This comes from an assumption that if something is already joined on the query level we will want to preload it anyways, eg. if you search over messages by the name of a user that sent the message you would not just join to filter but show user profile in the UI and so on (the cost for such decision is wasted IO when we don't render it).

**Comment by josevalim** [comment]
Ah, that’s a great example and it shows that we should go with `with_named_binding`. :) thanks!

PRs welcome!

**Comment by AndrewDryga** [comment]
On it :)

**Comment by AndrewDryga** [comment]
@josevalim on a side note are we open to allowing variables in bindings? This would further allow reducing code duplication with new `with_named_binding/3` usage patterns. Here is an example of how it could be used:

```elixir
  def with_preloaded_transport(queryable, parent_binding \\ :prospect_channel, binding \\ :transport) do
    with_named_binding(queryable, binding, fn queryable ->
      queryable
      |> join(
        :inner,
        [{^parent_binding, prospect_channel}],
        transport in assoc(prospect_channel, ^binding),
        as: ^binding
      )
    end)
  end
```

(I can provide more real world example from our codebase if needed.)

I did some experiments with it a few years ago but figured out it would be not a trivial change and not sure if Ecto wants to accept that. I'm not even sure that it's something we can implement without introducing breaking changes.

**Comment by v0idpwn** [comment]
I'm not a big fan of the proposal: the function is trivially written without using any internal API and in my experience the pattern isn't **that** common. 

Another thing I'm not a big fan is that the anonymous function returning an arbitrary query is very flexible and we can't really guarantee that the binding was added, yet we imply this in the function name.

Anyway, assuming we are settled, I have some concerns regarding the API.

### Design 

The issue describes the following:
```elixir
 with_named_binding(query, :comments, fn ->
    join(query, :left, c in assoc(p, :comments), as: :comments)
  end)
```

This probably isn't really good because it can't be possibly used within a pipeline:
```elixir
from(A)
|> with_named_binding(fn -> ??? end)
```
I assume it was a typo.

#3974 adds:
```elixir
 with_named_binding(query, :comments, fn  query->
    join(query, :left, c in assoc(p, :comments), as: :comments)
  end)
```

Which is better, but you have to write the binding name `:comments` two times and you can't really reuse the function for different names.

There's a third option which I think is better than both:
```elixir
 with_named_binding(query, :comments, fn  query, name->
    join(query, :left, c in assoc(p, :comments), as: name)
  end)
```

**Comment by fuelen** [comment]
What about building this semantic into the query? Like
```elixir
join(query, :left, c in assoc(p, :comments), as: :comments, on_binding_conflict: :nothing)
join(query, :left, c in assoc(p, :comments), as: :comments, on_binding_conflict: :raise) # default
```
----
Feedback from my experience: usually when we had usage of `has_named_binding?` it was a conditional applying of filters. We ended up with this library https://hexdocs.pm/composite/Composite.html which addresses the issue in the following way (example from the lib):
```elixir
User
|> where(active: true)
|> Composite.new(params)
|> Composite.param(:org_id, &filter_by_org_id/2)
|> Composite.param(:search_query, &search_by_full_name/2)
|> Composite.param(:org_name, &filter_by_org_name/2, requires: :org)
|> Composite.param(:org_type, &filter_by_org_type/2, requires: :org)
|> Composite.dependency(:org, &join_orgs/1)
|> Repo.all()
```
`join_orgs/1` is invoked only once and only if values in `:org_name` or in `:org_type` are present.
As a result, `filter_*` functions are a bit smaller, as they don't join tables inside.

**Comment by josevalim** [comment]
@fuelen I like that a lot. We could perhaps make it part of Ecto as:

    join(..., optional_as: :bar)

And then it will only be part of a query if someone uses it as a named binding. I am not sure about the complexity of implementing this though. We would also need to make so this join is not positional, until used, which can also be extra complexity.

The other option is your approach, where we specify the join at the end and it is only added if actually used.

**Comment by AndrewDryga** [comment]
I'm not a big fan of `on_binding_conflict: :nothing` approach. 

It feels like instead of having explicit code with functions we are adding more and more options (plus making implementation in Ecto more complex), options are easy to miss when you read the code (eg. on code reviews) and easier to forget when you write a lot of code that looks similar.

---

We have something similar to the @fuelen filtering approach too.

Our Query modules define a function required by a behavior:
```elixir
  @impl true
  def list_filters(_current_filters) do
    [
      %Filter{
        type: :string,
        field: :assigned_to_role,
        title: "Assigned To role",
        query_fun: &by_assigned_to_role_dynamic/2
        values: Roles.__values__()
      },
      ...
  end
```
The `Filter` struct holds enough information both for the back-end to apply the filter, to validate its value and for the front-end to render it nicely in the UI. The `_current_filters` is there for cases when depending on one of the filters new ones can become available.

The filters generally look something like this:
```elixir
 defp by_assigned_to_role_dynamic(queryable, role_name) do
    {with_joined_assigned_to(queryable), dynamic([assigned_to: assigned_to], assigned_to.role == ^role)}
  end
```
The tuple is returned here because we both might sometimes need to optionally preload something (leveraging `has_named_binding?/2`) and return a dynamic that actually filters the data.

The dynamics are used because we allow both simple filters, eg. `%{assigned_to_role: "role_name"}` and AST trees with and/or/not logic, eg. `{:or, [%{assigned_to_role: "role_name"}, %{for_everyone: true}]}`.

I plan to open-source our filtering soon-ish.

**Comment by elpikel** [comment]
Some time ago together with @zoldar we wrote macro that do just that. Maybe someone will be interested:
```elixir
  defmacro maybe_join(query, qual, binding \\ [], expr, opts \\ []) do
    named_binding =
      Enum.find(opts, fn
        {:as, _binding} ->
          true

        _ ->
          false
      end)

    case named_binding do
      {:as, binding_name} ->
        quote do
          query = unquote(query)

          if has_named_binding?(query, unquote(binding_name)) do
            query
          else
            join(query, unquote(qual), unquote(binding), unquote(expr), unquote(opts))
          end
        end

      _ ->
        raise ArgumentError, "Join must be named using `:as` option."
    end
  end
```

**Comment by josevalim** [comment]
Closing in favor of PR.
